### PR TITLE
✨ Improve condition in the controllers

### LINKED ISSUE #103: Have better conditions
/kind proposal

***Goals***
Improve conditions to improve the usability of the operator. 

**Detailed Description**

We should improve the conditions in CAPHV. There are the following tasks to do:

## Tasks
- [ ] Change rate limit condition to the same as in CAPH
- [ ] Add condition that no available device has been found
- [ ] printcolumns with message and reason like in CAPH
- [ ] conditions Summary like in CAPH
- [ ] Add condition for SSH key not found
- [ ] Add condition for server off
- [ ] Add conditions for TargetClusterReadyCondition
- [ ] Add condition for wrong API token

### PR DESCRIPTION
**Pull request by aniruddha2000**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part-of #103 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: controllers/hivelocitycluster_controller.go [hunk 1]
```diff
@@ -127,7 +127,7 @@ func (r *HivelocityClusterReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	secretManager := secretutil.NewSecretManager(log, r.Client, r.APIReader)
 	apiKey, hvSecret, err := getAndValidateHivelocityAPIKey(ctx, req.Namespace, hvCluster, secretManager)
 	if err != nil {
-		return hvAPIKeyErrorResult(ctx, err, hvCluster, infrav1.HivelocityClusterReady, r.Client)
+		return hvAPIKeyErrorResult(ctx, err, hvCluster, infrav1.HivelocityClusterReadyCondition, r.Client)
```
Comment (janiskemper) [inline controllers/hivelocitycluster_controller.go:130]: this should be CredentialsAvailableCondition no?

File: controllers/hivelocitymachine_controller.go [hunk 2]
```diff
@@ -120,8 +121,10 @@ func (r *HivelocityMachineReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	secretManager := secretutil.NewSecretManager(log, r.Client, r.APIReader)
 	hvAPIKey, _, err := getAndValidateHivelocityAPIKey(ctx, req.Namespace, hvCluster, secretManager)
 	if err != nil {
+		conditions.MarkFalse(hvCluster, infrav1.CredentialsAvailableCondition, infrav1.HivelocityWrongAPIKeyReason, clusterv1.ConditionSeverityError, err.Error())
 		return hvAPIKeyErrorResult(ctx, err, hivelocityMachine, infrav1.DeviceReadyCondition, r.Client)
 	}
+	conditions.MarkTrue(hvCluster, infrav1.CredentialsAvailableCondition)
```
Comment (janiskemper) [inline controllers/hivelocitymachine_controller.go:127]: this is too early to set it on true right? You set it a second time to true in device.go btw...

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: controllers/hivelocitycluster_controller.go (hunk 1 above)
Comment (aniruddha2000) [reply controllers/hivelocitycluster_controller.go:130]: right!

File: controllers/hivelocitymachine_controller.go (hunk 2 above)
Comment (aniruddha2000) [reply controllers/hivelocitymachine_controller.go:127]: right, first we check the API key the SSH key and if we found both then make true

File: controllers/hivelocitycluster_controller.go (hunk 1 above)
Comment (aniruddha2000) [reply controllers/hivelocitycluster_controller.go]: then have to think where to make the condition true for the cluster

Comment (aniruddha2000) [reply controllers/hivelocitycluster_controller.go]: https://github.com/hivelocity/cluster-api-provider-hivelocity/blob/c5be98c799432d14aba293b94fde3717a26d6d64/controllers/hivelocitycluster_controller.go#L374 This function we deal with a lot of secrets, do we need any credential available condition?

Comment (janiskemper) [reply controllers/hivelocitycluster_controller.go]: just look in CAPH, where we have the same reconcileTargetSecret and reconcileTargetManager. This is not related to the other credentials.

Comment (janiskemper) [reply controllers/hivelocitycluster_controller.go]: > then have to think where to make the condition true for the cluster

I am not sure whether we need it at all in the end.

**Comment by janiskemper** [comment]
what's the status @aniruddha2000 ?

**Comment by aniruddha2000** [comment]
@janiskemper you have a look now

**Review by janiskemper — CHANGES_REQUESTED** [review]
can you add the print columns "message" and "reason" to HivelocityMachine and HivelocityCluster?

**Comment by aniruddha2000** [comment]
Those changes are in the https://github.com/hivelocity/cluster-api-provider-hivelocity/pull/105/files#diff-8f714957aa724fa761ff2a8d1d2db28e542cc1f41b04e4f5b23be234b624a730

**Review by janiskemper — APPROVED** [review]

File: api/v1alpha1/conditions_const.go [hunk 3]
```diff
@@ -36,17 +36,28 @@ const (
 )
 
 const (
-	// HivelocityClusterReady reports on whether the Hivelocity cluster is in ready state.
-	HivelocityClusterReady clusterv1.ConditionType = "HivelocityClusterReady"
+	// HivelocityClusterReadyCondition reports on whether the Hivelocity cluster is in ready state.
+	HivelocityClusterReadyCondition clusterv1.ConditionType = "HivelocityClusterReady"
```
Comment (janiskemper) [inline api/v1alpha1/conditions_const.go:40]: do we still use this? The two reasons also sound related to CredentialsAvailable. Can you check?

Comment (aniruddha2000) [reply api/v1alpha1/conditions_const.go:40]: https://github.com/hivelocity/cluster-api-provider-hivelocity/blob/1b3e55ef94fdf8c74ca77f9ccc3e00bd9bd31e3a/controllers/hivelocitycluster_controller.go#L210 it is only used here, as a `MarkTrue()`, also the reasons associated with it are used by a `CredentialsAvailableCondition` MarkFalse(), what do you suggest?

Comment (janiskemper) [reply api/v1alpha1/conditions_const.go:40]: remove all of it

**Review by janiskemper — APPROVED** [review]

**Review by janiskemper — CHANGES_REQUESTED** [review]
we get an error when doing clusterscope.ClientConfig because the secret is not there yet. 

We should do the following error handling that is also in CAPH. Can you include it?

````
clientConfig, err := clusterScope.ClientConfig(ctx)
	if err != nil {
		if apierrors.IsNotFound(err) {
			conditions.MarkFalse(
				hetznerCluster,
				infrav1.TargetClusterReadyCondition,
				infrav1.KubeConfigNotFoundReason,
				clusterv1.ConditionSeverityInfo,
				"kubeconfig not found (yet)",
			)
			return nil, nil
		}
````

**Comment by aniruddha2000** [comment]
https://github.com/hivelocity/cluster-api-provider-hivelocity/blob/9956a85a2f1dab39ff081ea3a8c7d7f4cf4cd8ea/controllers/hivelocitycluster_controller.go#L570 i have included that in the other PR

**Review by janiskemper — APPROVED** [review]

**Pull request merged**